### PR TITLE
Add more config parameters to extensions

### PR DESCRIPTION
**Pull request by ss2165**

### CODE REVIEW

**Comment by cqc-alec** [comment]
I'm trying to run all the tests, including those that need credentials using this branch. (I still can't run the azure test, sadly.) I'm left with a small number of failures:
* braket backend_test test_rigetti_with_rerouting fails with an error about a disconnected architecture graph. This seems like a bug in tket, so I'll raise that as a separate issue.
* honeywell -- when I try to run the honeywell tests they all fail with authentication errors: "pytket.extensions.honeywell.backends.api_wrappers.HQSAPIError: HTTP error while logging in: 403". My config file for honeywell contains my username (email address), as before; do I need to change something here?
* qiskit qiskit_convert_test.py test_tketautopass is failing in the line `tk_backend = IBMQBackend(backend.name())` in `tket_pass.py`, because "More than one provider matches the specified criteria". This seems like a bug -- the name isn't necessarily enough to specify the backend. Yet for some reason it doesn't happen to me on develop. Still trying to understand this.

**Comment by cqc-alec** [comment]
The honeywell errors happen on develop too so I guess not related to the config changes.

**Comment by cqc-alec** [comment]
Re qiskit issue: this code:
```
from qiskit import IBMQ
from pytket.extensions.qiskit.tket_pass import TketAutoPass

IBMQ.load_account()
provider = IBMQ.providers(hub="ibm-q")[0]
back = provider.get_backend("ibmq_santiago")
tkpass = TketAutoPass(back, 0)
```
fails on this branch ("More than one provider matches the specified criteria") but succeeds on develop. I still don't understand this, because it doesn't seem to be dependent on pytket at all.

**Comment by cqc-alec** [comment]
Honeywell: on both develop and this branh, when I do:
```
from pytket.extensions.honeywell import HoneywellBackend
b = HoneywellBackend()
```
I get
```
Your id token is expired. Refreshing...
[...]
HQSAPIError: HTTP error while logging in: 403
```
Maybe a problem at their end?

File: modules/pytket-aqt/pytket/extensions/aqt/backends/aqt.py [hunk 1]
```diff
@@ -88,22 +96,36 @@ class AQTBackend(Backend):
...
         self._label = label
+        config = AQTConfig.from_default_config_file()
+
+        if access_token is None:
+            access_token = config.access_token
+        if access_token is None:
+            raise AqtAuthenticationError()
+
         self._header = {"Ocp-Apim-Subscription-Key": access_token, "SDK": "pytket"}
+        print(self._header)
```
Comment (cqc-alec) [inline modules/pytket-aqt/pytket/extensions/aqt/backends/aqt.py:128]: Remove.

File: modules/pytket-braket/pytket/extensions/braket/backends/braket.py [hunk 2]
```diff
@@ -183,25 +185,47 @@ class BraketBackend(Backend):
...
+        s3_folder: Optional[str] = None,
+        device_type: Optional[str] = None,
+        provider: Optional[str] = None,
     ):
         """
         Construct a new braket backend.
 
         If `local=True`, other parameters are ignored.
 
+        All parameters except `device` can be set in config using
```
Comment (cqc-alec) [inline modules/pytket-braket/pytket/extensions/braket/backends/braket.py:199]: Why not "device"?

Comment (cqc-alec) [inline modules/pytket-braket/pytket/extensions/braket/backends/braket.py:199]: Docs should state what the defaults are if not specified or set in config.

File: modules/pytket-braket/pytket/extensions/braket/backends/braket.py [hunk 3]
```diff
@@ -183,25 +185,47 @@ class BraketBackend(Backend):
...
+        if s3_bucket is None:
+            s3_bucket = config.s3_bucket
+        if s3_folder is None:
+            s3_bucket = config.s3_folder
+        if device_type is None:
+            s3_bucket = config.device_type
+        if provider is None:
+            s3_bucket = config.provider
+
+        # set defaults if not overriden
```
Comment (cqc-alec) [inline modules/pytket-braket/pytket/extensions/braket/backends/braket.py:221]:
```suggestion
        # set defaults if not overridden
```

File: modules/pytket-braket/pytket/extensions/braket/backends/config.py [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+
+
+def set_braket_config(
+    s3_bucket: Optional[str] = None,
+    s3_folder: Optional[str] = None,
+    device_type: Optional[str] = None,
+    provider: Optional[str] = None,
+) -> None:
+    """Set default values for any of s3_bucket, s3_folder, device_type or provider
+    for AWS Braket. Can be overriden in backend construction."""
```
Comment (cqc-alec) [inline modules/pytket-braket/pytket/extensions/braket/backends/config.py:50]:
```suggestion
    for AWS Braket. Can be overridden in backend construction."""
```

File: modules/pytket-ionq/pytket/extensions/ionq/backends/ionq.py [hunk 5]
```diff
@@ -62,35 +63,53 @@
...
+    def __init__(self) -> None:
+        super().__init__("No IonQ api key provided or found in config file.")
+
+
 class IonQBackend(Backend):
     """
     Interface to an IonQ device.
+
+    Requires a valid API key/access token, this can either be provided as a
+    parameter or set in config using :py:meth:`pytket.extensions.aqt.set_aqt_config`
```
Comment (cqc-alec) [inline modules/pytket-ionq/pytket/extensions/ionq/backends/ionq.py:78]:
```suggestion
    parameter or set in config using :py:meth:`pytket.extensions.ionq.set_ionq_config`
```

File: modules/pytket-qiskit/pytket/extensions/qiskit/backends/config.py [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+
+
+def set_ibmq_config(
+    hub: Optional[str] = None,
+    group: Optional[str] = None,
+    project: Optional[str] = None,
+    ibmq_api_token: Optional[str] = None,
+) -> None:
+    """Set default values for any of hub, group, project or API token
+    for your IBMQ provider. Can be overriden in backend construction."""
```
Comment (cqc-alec) [inline modules/pytket-qiskit/pytket/extensions/qiskit/backends/config.py:50]:
```suggestion
    for your IBMQ provider. Can be overridden in backend construction."""
```

File: modules/pytket-qsharp/pytket/extensions/qsharp/backends/config.py [hunk 7]
```diff
@@ -12,26 +12,44 @@
...
+        )
+
+
+def set_azure_config(
+    resourceId: Optional[str] = None,
+    location: Optional[str] = None,
+    storage: Optional[str] = None,
+) -> None:
+    """Set default values for any of resourceId, location or storage
+    for your Azure Quantum Workspace. Can be overriden in backend construction."""
```
Comment (cqc-alec) [inline modules/pytket-qsharp/pytket/extensions/qsharp/backends/config.py:47]:
```suggestion
    for your Azure Quantum Workspace. Can be overridden in backend construction."""
```

File: docs/build-docs [hunk 8]
```diff
@@ -134,11 +134,10 @@ if __name__ == "__main__":
             shutil.copytree(
                 MODULES_DIR / f"pytket-{module}" / "docs" / "build",
                 dest / module,
-                dirs_exist_ok=True,
```
[the remark this answers is not in the document]
Comment (ss2165) [reply docs/build-docs:137]: no, sorry

File: modules/pytket-braket/pytket/extensions/braket/backends/braket.py (hunk 2 above)
Comment (ss2165) [reply modules/pytket-braket/pytket/extensions/braket/backends/braket.py:199]: just to match the existing interfaces of other backends where the device is always specified at code-time. Willing to change this later

**Comment by cqc-alec** [comment]
> Honeywell: on both develop and this branh, when I do:
[...]
> Maybe a problem at their end?

Seems to be a problem at my end; happy to ignore for now.

File: modules/pytket-aqt/pytket/extensions/aqt/backends/aqt.py (hunk 1 above)
Comment (ss2165) [reply modules/pytket-aqt/pytket/extensions/aqt/backends/aqt.py]: gone

File: modules/pytket-braket/pytket/extensions/braket/backends/braket.py (hunk 2 above)
Comment (ss2165) [reply modules/pytket-braket/pytket/extensions/braket/backends/braket.py:199]: done

File: modules/pytket-qsharp/pytket/extensions/qsharp/backends/config.py (hunk 7 above)
Comment (ss2165) [reply modules/pytket-qsharp/pytket/extensions/qsharp/backends/config.py]: done

**Review by cqc-alec — APPROVED** [review]

**Comment by ss2165** [comment]
> 
> 
> Re qiskit issue: this code:
> 
> ```
> from qiskit import IBMQ
> from pytket.extensions.qiskit.tket_pass import TketAutoPass
> 
> IBMQ.load_account()
> provider = IBMQ.providers(hub="ibm-q")[0]
> back = provider.get_backend("ibmq_santiago")
> tkpass = TketAutoPass(back, 0)
> ```
> 
> fails on this branch ("More than one provider matches the specified criteria") but succeeds on develop. I still don't understand this, because it doesn't seem to be dependent on pytket at all.

There was a bug when no provider parameters were specified, should be fixed with latest commit

**Review by cqc-alec — APPROVED** [review]

**Pull request merged**